### PR TITLE
Support for Signet

### PR DESCRIPTION
**Pull request by dr-orlovsky**
This is dependent on #238 and #240 and includes commit history from them (w/o this it would not build). Once these PRs will be merged I will rebase on master and clear the history so this PR can be reviewed (it has quite a few changes).

The only relevant commit in the history to Signet support is <https://github.com/romanz/electrs/commit/1a4fc7851ccb276614798c296339cd98e14bfa3a>

### CODE REVIEW

**Comment by romanz** [comment]
LGTM, many thanks for the contribution!
Will be happy to merge when rust-bitcoin 0.24 is released.

**Comment by Kixunil** [comment]
Would rebasing the changes make them more clear now?

**Comment by dr-orlovsky** [comment]
Rebased, now the history is clear. Waiting for bitcoin 0.24

**Comment by dr-orlovsky** [comment]
Now, with the new rust-bitcoin version release I've updated dependency PR #240 and rebased onto it. However, to merge this we still have to wait for https://github.com/rust-bitcoin/rust-bitcoin/pull/291 to be merged into rust-bitcoin and yet another version release after that

**Comment by dr-orlovsky** [comment]
Docker container & image working with signet + whole signet package (bitcoind+electrs+lightningd):
- https://github.com/LNP-BP/docker/blob/master/electrs-signet/Dockerfile
- https://hub.docker.com/r/lnpbp/electrs/signet
- https://github.com/LNP-BP/docker/blob/master/signet/docker-compose.yml

**Comment by romanz** [comment]
CC: @kallewoof
Cherry-picked over (WIP) `next` branch - fac0da4 seems to be working :)
```
$ ./query.sh signet tb1q0k8lrhuqxj32rehjg7g2sxnegrzhqawx8xkadd
[2020-11-05T14:07:36.815Z INFO  electrs_index::metrics] serving Prometheus metrics on 127.0.0.1:34224
[2020-11-05T14:07:36.815Z DEBUG electrs_index::db] opening DB with Options { path: "./db/signet", bulk_import: true, low_memory: false }
[2020-11-05T14:07:37.573Z DEBUG electrs_index::db] DB Config { compacted: false, format: 1 }
[2020-11-05T14:07:37.628Z DEBUG electrs_index::index] read 0 blocks (0.000 MB) from DB at 0 ms
[2020-11-05T14:07:37.628Z DEBUG electrs_index::map] added 0/0 headers at 0 ms
[2020-11-05T14:07:37.710Z DEBUG electrs_index::index] loaded 10433 block locations (10433 new) at 81 ms
[2020-11-05T14:07:37.711Z INFO  electrs_index::index] reading 10433 new blocks from 1 blk*.dat files
[2020-11-05T14:07:37.767Z DEBUG electrs_index::index] indexer #1: indexed 10433 blocks from file #0 into 24075 rows
[2020-11-05T14:07:37.774Z DEBUG electrs_index::index] writing index 24075 rows from 10433 blocks
[2020-11-05T14:07:37.791Z INFO  electrs_index::index] indexed 10433 new blocks, 24075 DB rows, took 0.163s
[2020-11-05T14:07:38.397Z INFO  electrs_index::db] starting config compaction
[2020-11-05T14:07:38.452Z INFO  electrs_index::db] starting headers compaction
[2020-11-05T14:07:38.497Z INFO  electrs_index::db] starting txid compaction
[2020-11-05T14:07:38.553Z INFO  electrs_index::db] starting index compaction
[2020-11-05T14:07:38.653Z INFO  electrs_index::db] finished full compaction
[2020-11-05T14:07:38.910Z INFO  electrs_index::db] auto-compactions enabled
[2020-11-05T14:07:38.924Z DEBUG electrs_index::map] added 10433/10433 headers at 13 ms
[2020-11-05T14:07:38.925Z INFO  electrs_index::map] verified 10433 blocks, tip=000000b443529edefcd6c7ba398a16ad037abd71a89f74ebedc5a9459f907975 @ 2020-11-05T14:00:00+00:00
[2020-11-05T14:07:38.925Z DEBUG electrs_index::index] 4e6eb8f2a2659a8529ecf88148cdd58d368690397610c04f85d59da4dcef97a6 has 1 txid rows
[2020-11-05T14:07:38.925Z DEBUG electrs_query] received: 10.00000000 @ 1 outputs (scanned 1 txs, 2 outputs)
[2020-11-05T14:07:38.925Z DEBUG electrs_query] spent:    0.00000000 @ 0 inputs (scanned 1 txs, 1 inputs)
[2020-11-05T14:07:38.925Z INFO  electrs_query] 71db759c6747a99e5e958c25aa6fad7ac12ae00c8ec3414587d9ed454c85118a:1              10.00000000 @ 10409 tb1q0k8lrhuqxj32rehjg7g2sxnegrzhqawx8xkadd
[2020-11-05T14:07:38.925Z INFO  electrs_query] total: 10.00000000 BTC
```

File: Cargo.toml
Comment (kallewoof) [inline Cargo.toml:27]: Rust-bitcoin has merged signet support so this should use the upstream version. See https://github.com/rust-bitcoin/rust-bitcoin/pull/291

Comment (kallewoof) [reply Cargo.toml:27]: I realized this was cherry picked in a different PR, ignore above.

**Comment by kallewoof** [comment]
> CC: @kallewoof
> Cherry-picked over [the WIP `next` branch](https://github.com/romanz/electrs/issues/308) - [fac0da4](https://github.com/romanz/electrs/commit/fac0da4577709f7bb1bd085ebd59bae0c9ffe930) seems to be working :)

utACK https://github.com/romanz/electrs/commit/fac0da4577709f7bb1bd085ebd59bae0c9ffe930

**Comment by Kixunil** [comment]
What's blocking the merge of this? Would be nice to bump `bitcoin` version so that `secp256k1` can be made optional cleanly when my PR gets merged. :)

**Comment by romanz** [comment]
It is now merged in `next` [branch](https://github.com/romanz/electrs/issues/308).

**Comment by Kixunil** [comment]
What about merging to master? Or do you plan for `next` to be the next release?

**Comment by dr-orlovsky** [comment]
Well, this require new signet-supporting rust-bitcoin version to be released; which has not happened yet. I monitor the situation nearly daily.

**Comment by Kixunil** [comment]
Hmm, the constants are already there. At least that was the problem when attempting to try `electrs` with my patch to remove `secp256k1`.

**Comment by dr-orlovsky** [comment]
They are in the master, but not in crates.io (pending release)

**Comment by romanz** [comment]
> What about merging to master? Or do you plan for next to be the next release?

I am planning to port the missing features from `master` to `next` - e.g. configuration, better error handling - and then merge it into `master` as part of `0.9.*` release, since it's DB schema is not backwards-compatible.
I plan to support the `0.8.*` branch for the near future, but I prefer to continue the development using the new code base.

**Comment by dr-orlovsky** [comment]
With rust-bitcoin 0.26 release this is finally ready to be merged into master

**Review by romanz — APPROVED** [review]
Many thanks!

**Comment by dr-orlovsky** [comment]
@romanz can I pls ask you to publish a release (tag+crate), so I can update my docker images?

**Comment by romanz** [comment]
[Tagged](https://github.com/romanz/electrs/releases/tag/v0.8.7) and [released](https://crates.io/crates/electrs/0.8.7) :)

**Pull request merged**
